example/src/main/java/io/netty/example/tunechoserver/InetUtils.java

```diff
@@ -0,0 +1,329 @@
+            }
+
+            // Get the protocol.
+            ipHeader.protocol = packetBuf.getUnsignedByte(ipHeaderOffset + IPV6_HEADER_OFFSET_NEXT_HEADER);
+
+            // Get the source and destination addresses.
+            ipHeader.sourceAddress = new byte[16];
+            ipHeader.destAddress = new byte[16];
+            packetBuf.getBytes(ipHeaderOffset + InetUtils.IPV6_HEADER_OFFSET_SOURCE_ADDR, ipHeader.sourceAddress);
+            packetBuf.getBytes(ipHeaderOffset + InetUtils.IPV6_HEADER_OFFSET_DEST_ADDR, ipHeader.destAddress);
+
+        // Otherwise if the packet is an IPv6 packet...
+        } else if (ipHeader.ipVersion == IP_VERSION_4) {
+
+            // Get the length of the header.
+            ipHeader.headerLength =
+                    (packetBuf.getUnsignedByte(ipHeaderOffset + IPV4_HEADER_OFFSET_HEADER_LENGTH) & 0x0F) * 4;
+
+            // Fail if header length value is too small.
+            if (ipHeader.headerLength < IPV4_MIN_HEADER_LENGTH) {
+                throw new IllegalArgumentException("Invalid IPv4 header length");
+            }
+
+            // Fail if the packet is too small to contain the IPv4 header.
+            if (packetBuf.readableBytes() < ipHeaderOffset + ipHeader.headerLength) {
+                throw new IllegalArgumentException("IP packet too small");
+            }
+
+            // Get the total length of the packet.
+            int totalLength = packetBuf.getUnsignedShort(ipHeaderOffset + IPV4_HEADER_OFFSET_TOTAL_LENGTH);
+
+            // Fail if the packet is smaller than the total length.
+            if (packetBuf.readableBytes() < ipHeaderOffset + totalLength) {
+                throw new IllegalArgumentException("IP packet too small");
+            }
+
+            // Compute the payload length
+            ipHeader.payloadLength = totalLength - ipHeader.headerLength;
+
+            // Get the protocol.
+            ipHeader.protocol = packetBuf.getUnsignedByte(ipHeaderOffset + IPV4_HEADER_OFFSET_PROTOCOL);
+
+            // Get the source and destination addresses.
+            ipHeader.sourceAddress = new byte[4];
+            ipHeader.destAddress = new byte[4];
+            packetBuf.getBytes(ipHeaderOffset + InetUtils.IPV4_HEADER_OFFSET_SOURCE_ADDR, ipHeader.sourceAddress);
+            packetBuf.getBytes(ipHeaderOffset + InetUtils.IPV4_HEADER_OFFSET_DEST_ADDR, ipHeader.destAddress);
+
+            // Get IP checksum.
+            ipHeader.checksum = packetBuf.getUnsignedShort(ipHeaderOffset + IPV4_HEADER_OFFSET_CHECKSUM);
+
+        // Otherwise the packet type is unknown.
+        } else {
+            throw new UnsupportedOperationException(
+                    "IP version unsupported: " + ipVersionToString(ipHeader.ipVersion));
+        }
+
+        return ipHeader;
+    }
+
+    public static final class ICMPHeader {
+        public int type;
+        public int code;
+        public int checksum;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make checksum a static final constant or non-public and provide accessors if needed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)